packages/react-devtools-shared/src/backend/fiber/renderer.js

```diff
@@ -2348,7 +2397,12 @@ export function attach(
       ownerInstance.source = componentInfo.debugStack;
     }
     const ownerID = ownerInstance === null ? 0 : ownerInstance.id;
-    const parentID = parentInstance ? parentInstance.id : 0;
+    const parentID = parentInstance
+      ? parentInstance.kind === FILTERED_FIBER_INSTANCE
+        ? // A Filtered Fiber Instance will always have a Virtual Instance as a parent.
```

Comment:
Is this statement true only because of the current implementation (lines 2771-2772)?

Like if I've just filtered out some client component based on its name, then its not Virtual Instance only, right?

Reply:
Yes, it's only true based on the current implementation and if that changes then this, and any other place sending instruction to the frontend, would have to turn into a loop backtracking until it hit something unfiltered.

It's basically because all virtual instances themselves come from backtracking from a Fiber. It's impossible for a virtual instance to exist without bottoming out into a Fiber (for example a Server Component that renders null doesn't actually show up in the tree today but it could be modeled by an empty Fragment Fiber). They need some Fiber to find their place in the tree if we traverse filtered things like the Suspense/Error case or finding child Host Instances.

There's no need to do that for fiber instances since they already have a Fiber that we can walk and so therefore this is already a Virtual Instance.

Technically this holds true as long it's not two Filtered Fibers in a row. Once you could have two filtered fibers nested it would need to turn into a loop.